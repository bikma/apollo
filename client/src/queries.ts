import { gql } from 'apollo-boost'

export const GET_VEHICLES = gql`
  query Vehicles {
    vehicles {
      backImg
      comments
      driverCondition
      driverLicense
      driverName
      driverPhone
      entryCondition
      frontImg
      id
      inTime
      leftImg
      loadedCondition
      loadedWeight
      numberPlateImg
      outTime
      registrationNumber
      reportingTime
      rightImg
      safetyHelmet
      safetyInstructions
      safetyJacket
      unloadedCondition
      unloadedWeight
      vehicleType
      wheelsSecured
    }
  }
`
