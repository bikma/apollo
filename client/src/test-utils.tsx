import * as React from 'react'

import { act, cleanup, render, wait } from '@testing-library/react'

import { MockedProvider } from '@apollo/react-testing'
import { RenderResult } from '@testing-library/react'

interface It {
  component: JSX.Element
  mocks: Array<any>
  name: string
  wait?: number
  callback?: (response: RenderResult) => void
}

interface Suite {
  name: string
  its: Array<It>
}

export const runSuites = (suites: Array<Suite>) => {
  if (!suites) return
  suites.forEach((suite: Suite) => {
    describe(suite.name, () => {
      afterEach(cleanup)
      suite.its.forEach((_it: It) => {
        it(_it.name, async () => {
          await act(async () => {
            const response: RenderResult = render(
              <MockedProvider mocks={_it.mocks} addTypename={false}>
                {_it.component}
              </MockedProvider>
            )
            if (_it.wait !== 0)
              await waitUntilLoadingIsFinished(response.queryByText)
            expect(response.container).toMatchSnapshot()
            if (_it.callback) _it.callback(response)
          })
        })
      })
    })
  })
}

const waitUntilLoadingIsFinished = (queryByText: any) =>
  wait(() => {
    const isLoading = queryByText('Loading') != null
    expect(isLoading).toBe(false)
  })
