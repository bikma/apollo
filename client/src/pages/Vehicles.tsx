import { GET_VEHICLES } from '../queries'
import React from 'react'
import { Vehicles as VehicleType } from '../__generated__/types'
import { useQuery } from '@apollo/react-hooks'

export const Vehicles: React.FC = () => {
  const { loading, error, data } = useQuery<VehicleType>(GET_VEHICLES)
  if (loading) return <div data-testid="loading">Loading</div>
  if (error) return <div data-testid="error">{error.message}</div>
  if (!data) return <div data-testid="no-data">No data</div>
  const vehicles = data.vehicles
  return (
    <React.Fragment>
      {vehicles &&
        vehicles.map((v, i) => <p key={i}>{v && v.registrationNumber}</p>)}
    </React.Fragment>
  )
}

export default Vehicles
