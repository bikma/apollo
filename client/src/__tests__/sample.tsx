import '@testing-library/jest-dom/extend-expect'

import * as React from 'react'

import { GET_VEHICLES } from '../queries'
import { RenderResult } from '@testing-library/react'
import Vehicles from '../pages/Vehicles'
import { runSuites } from '../test-utils'

const request = {
  query: GET_VEHICLES
}

const mocksWithSuccess = [
  {
    request,
    result: {
      data: {
        vehicles: [
          {
            backImg: 'example',
            comments: 'example',
            driverCondition: 12345,
            driverLicense: '1234',
            driverName: 'example',
            driverPhone: 12345,
            entryCondition: 12345,
            frontImg: 'example',
            id: 1,
            inTime: 1568737336,
            leftImg: 'example',
            loadedCondition: 12345,
            loadedWeight: 12345,
            numberPlateImg: 'example',
            outTime: 1568737336,
            registrationNumber: 'example',
            reportingTime: 1568737336,
            rightImg: 'example',
            safetyHelmet: 12345,
            safetyInstructions: 12345,
            safetyJacket: 12345,
            unloadedCondition: 12345,
            unloadedWeight: 12345,
            vehicleType: 'example',
            wheelsSecured: 12345
          }
        ]
      }
    }
  }
]
const mocksWithEmpty = [
  {
    request,
    result: {
      data: {
        vehicles: [null]
      }
    }
  }
]
const mocksWithError = [
  {
    error: new Error('Something went wrong'),
    request
  }
]

const loading = (response: RenderResult) => {
  // additional tests apart from snapshot match
  expect(response.getByTestId('loading')).toHaveTextContent('Loading')
}

const suites = [
  {
    its: [
      {
        callback: loading,
        component: <Vehicles />,
        mocks: [],
        name: 'renders loading',
        wait: 0
      },
      {
        component: <Vehicles />,
        mocks: mocksWithSuccess,
        name: 'renders data'
      },
      {
        component: <Vehicles />,
        mocks: mocksWithError,
        name: 'renders error'
      },
      {
        component: <Vehicles />,
        mocks: mocksWithEmpty,
        name: 'renders empty'
      }
    ],
    name: 'App'
  }
]

runSuites(suites)
