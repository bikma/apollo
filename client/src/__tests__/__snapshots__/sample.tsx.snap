// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders data 1`] = `
<div>
  <p>
    example
  </p>
</div>
`;

exports[`App renders empty 1`] = `
<div>
  <p />
</div>
`;

exports[`App renders error 1`] = `
<div>
  <div
    data-testid="error"
  >
    Network error: Something went wrong
  </div>
</div>
`;

exports[`App renders loading 1`] = `
<div>
  <div
    data-testid="loading"
  >
    Loading
  </div>
</div>
`;
